node/src/components/block_validator.rs

```diff
@@ -59,114 +68,114 @@ pub(crate) struct BlockValidationState<T> {
-                                    | FetchResult::FromPeer(deploy, _) => {
-                                        if deploy.header().timestamp() > block_timestamp {
-                                            Event::DeployMissing(dh_found)
-                                        } else {
-                                            Event::DeployFound(dh_found)
-                                        }
-                                    }
-                                },
-                                move || Event::DeployMissing(dh_not_found),
-                            )
-                    })
-                    .collect();
+                    .map(|deploy_hash| **deploy_hash)
+                    .collect::<HashSet<_>>();
+                if block_deploys.is_empty() {
+                    // If there are no deploys, return early.
+                    return responder.respond((true, block)).ignore();
+                }
 
                 // TODO: Clean this up to use `or_insert_with_key` once
                 // https://github.com/rust-lang/rust/issues/71024 is stabilized.
                 match self.validation_states.entry(block) {
                     Entry::Occupied(mut entry) => {
-                        // The entry already exists. We register ourselves as someone interested in
-                        // the ultimate validation result.
-                        entry.get_mut().responders.push(responder);
+                        // The entry already exists.
+                        if entry.get().missing_deploys.is_empty() {
+                            // Block has already been validated successfully, early return to
+                            // caller.
+                            effects.extend(responder.respond((true, entry.key().clone())).ignore());
+                        } else {
+                            // We register ourselves as someone interested in the ultimate
+                            // validation result.
+                            entry.get_mut().responders.push(responder);
+                        }
                     }
                     Entry::Vacant(entry) => {
                         // Our entry is vacant - create an entry to track the state.
                         let missing_deploys: HashSet<DeployHash> =
                             entry.key().deploys().iter().map(|hash| **hash).collect();
 
+                        let in_flight = &mut self.in_flight;
+                        let chainspec = Arc::clone(&self.chainspec);
+                        let fetch_effects: Effects<Event<T, I>> = block_deploys
+                            .iter()
+                            .flat_map(|deploy_hash| {
+                                let chainspec = Arc::clone(&chainspec);
+                                // For every request, increase the number of in-flight...
+                                in_flight.inc(deploy_hash);
```

Comment:
`in_flight` uses a `HashMap` internally, this might end up doing a lot of `hash`-ing when we increment the number of in-flight requests for every deploy in a block.